File: web.cpp
#include <WiFi.h>
#include <WiFiClient.h>
#include <WebServer.h>
#include <ESPmDNS.h>
#include <esp_heap_caps.h>
#include <ArduinoJson.h>
#include <freertos/FreeRTOS.h>
#include <freertos/task.h>

#include "FS.h"
#include "SPIFFS.h"

static WebServer server(80);

void storeConfig();

void handleRoot()
{
    server.send(200, "text/plain", "hello from esp32!");
}


void handleSettings()
{
    File file = SPIFFS.open("/config.html", "r");
    if (!file)
    {
        Serial.println("file error");
    }
    server.streamFile(file, "text/html");
    file.close();
}


void handleConfig()
{
    storeConfig();

    // server.send(302, "text/html");
    File htmlfile = SPIFFS.open("/config.html", "r");
    if (!htmlfile)
    {
        Serial.println("file error");
    }
    server.streamFile(htmlfile, "text/html");
    htmlfile.close();
}


void handleNotFound()
{
    String message = "File Not Found\n\n";
    message += "URI: ";
    message += server.uri();
    message += "\nMethod: ";
    message += (server.method() == HTTP_GET) ? "GET" : "POST";
    message += "\nArguments: ";
    message += server.args();
    message += "\n";
    for (uint8_t i = 0; i < server.args(); i++)
    {
        message += " " + server.argName(i) + ": " + server.arg(i) + "\n";
    }
    server.send(404, "text/plain", message);
}


void setupAP()
{
    Serial.println("Configuring access point...");
    uint8_t mac[6];
    char buff[128];
    WiFi.macAddress(mac);
    sprintf(buff, "T-EPD47-%02X%02X", mac[4], mac[5]);
    WiFi.softAP(buff);
    Serial.printf("The hotspot has been established");
    Serial.printf("please connect to the %s and output 192.168.4.1/settings in the browser to access the data page \n", buff);
}


void webTask(void *args)
{
    while (1)
    {
        server.handleClient();
        delay(10); //allow the cpu to switch to other tasks
    }
    vTaskDelete(NULL);
}


void setupWEB(void)
{
    Serial.begin(115200);
    setupAP();

    server.on("/", handleRoot);
    server.on("/settings", handleSettings);
    server.on("/config", HTTP_POST, handleConfig);
    server.on("/config", HTTP_GET, []() {
        File file = SPIFFS.open("/config.json", "r");
        size_t sent = server.streamFile(file, "application/json");
        file.close();
        return;
    });
    server.on("/restart", HTTP_GET, []() {
        server.send(200);
        delay(1000);
        ESP.restart();
        return;
    });

    server.onNotFound(handleNotFound);

    server.begin();
    Serial.println("HTTP server started");

    TaskHandle_t t1;
    xTaskCreatePinnedToCore((void (*)(void *))webTask, "webTask", 8192, NULL, 10, &t1, 0);
}


void storeConfig()
{
    StaticJsonDocument<1024> doc;

    File configfile = SPIFFS.open("/config.json", "r");
    DeserializationError error = deserializeJson(doc, configfile);
    if (error)
        Serial.println(F("Failed to read file, using default configuration"));
    configfile.close();

    for ( uint8_t i = 0; i < server.args(); i++ )
    {
        if (server.argName(i).equals("ssid"))
        {
            doc["WLAN"]["ssid"] = server.arg(i);
        }
        else if (server.argName(i).equals("password"))
        {
            doc["WLAN"]["password"] = server.arg(i);
        }
        else if (server.argName(i).equals("apikey"))
        {
            doc["OpenWeather"]["apikey"] = server.arg(i);
        }
        else if (server.argName(i).equals("server"))
        {
            doc["OpenWeather"]["server"] = server.arg(i);
        }
        else if (server.argName(i).equals("country"))
        {
            doc["OpenWeather"]["country"] = server.arg(i);
        }
        else if (server.argName(i).equals("city"))
        {
            doc["OpenWeather"]["city"] = server.arg(i);
        }
        else if (server.argName(i).equals("hemisphere"))
        {
            doc["OpenWeather"]["hemisphere"] = server.arg(i);
        }
        else if (server.argName(i).equals("units"))
        {
            doc["OpenWeather"]["units"] = server.arg(i);
        }
        else if (server.argName(i).equals("ntp_server"))
        {
            doc["ntp"]["server"] = server.arg(i);
        }
        else if (server.argName(i).equals("ntp_timezone"))
        {
            doc["ntp"]["timezone"] = server.arg(i);
        }
        else if (server.argName(i).equals("on_time"))
        {
            doc["schedule_power"]["on_time"] = server.arg(i);
        }
        else if (server.argName(i).equals("off_time"))
        {
            doc["schedule_power"]["off_time"] = server.arg(i);
        }
    }

    configfile = SPIFFS.open("/config.json", FILE_WRITE);
    if (!configfile)
    {
        Serial.println("file error");
    }
    if (serializeJson(doc, configfile) == 0)
    {
        Serial.println(F("Failed to write to file"));
    }
    configfile.close();
}

File: micropython/web.py
import ure as re
import picoweb

try:
    import network
except:
    pass

try:
    import usocket as socket
except:
    import socket

from machine import reset
from time import sleep
from json import dump, load

def settings(req, resp):
    # You can construct an HTTP response completely yourself, having
    # a full control of headers sent...
    yield from picoweb.start_response(resp, content_type = "text/html")
    htmlFile = open('www/config.html', 'r')
    for line in htmlFile:
        yield from resp.awrite(line)
    htmlFile.close()


def restart(req, resp):
    if req.method == "GET":
        yield from picoweb.start_response(resp)
    sleep(1)
    reset()


def config(req, resp):
    if req.method == "GET":
        yield from picoweb.start_response(resp, content_type = "application/json")
        jsonfile = open('config.json', 'r')
        for line in jsonfile:
            yield from resp.awrite(line)
        jsonfile.close()
    elif req.method == "POST":
        yield from req.read_form_data()
        storeConfig(req)
        yield from picoweb.start_response(resp, content_type = "text/html")
        htmlFile = open('www/config.html', 'r')
        for line in htmlFile:
            yield from resp.awrite(line)
        htmlFile.close()

    # Here's how you extract matched groups from a regex URI match


ROUTES = [
    # You can specify exact URI string matches...
    ("/settings", settings),
    ("/restart", restart),
    ("/config", config),
]


def storeConfig(req):
    jsonfile = open('config.json', 'r')
    j = load(jsonfile)
    jsonfile.close()

    j["WLAN"]["ssid"] = req.form["ssid"]
    j["WLAN"]["password"] = req.form["password"]
    j["OpenWeather"]["apikey"] = req.form["apikey"]
    j["OpenWeather"]["server"] = req.form["server"]
    j["OpenWeather"]["country"] = req.form["country"]
    j["OpenWeather"]["city"] = req.form["city"]
    j["OpenWeather"]["hemisphere"] = req.form["hemisphere"]
    j["OpenWeather"]["units"] = req.form["units"]
    j["ntp"]["server"] = req.form["ntp_server"]
    j["ntp"]["timezone"] = req.form["ntp_timezone"]
    j["schedule_power"]["on_time"] = req.form["on_time"]
    j["schedule_power"]["off_time"] = req.form["off_time"]

    jsonfile = open('config.json', 'w')
    dump(j, jsonfile)
    jsonfile.close()
    del j


def setupAP():
    ap = network.WLAN(network.AP_IF) # create access-point interface
    mac = list(ap.config('mac'))
    ssid = 'EPD47-{:0>2X}{:0>2X}{:0>2X}'.format(mac[3], mac[4], mac[5])
    ap.config(essid=ssid) # set the ESSID of the access point
    ap.config(max_clients=5) # set how many clients can connect to the network
    ap.active(True)         # activate the interface
    print("The hotspot has been established")
    print("please connect to the {} and output 192.168.4.1/settings in the browser to access the data page", ssid)


def setupWEB():
    setupAP()
    app = picoweb.WebApp(__name__, ROUTES)
    app.run(host="0.0.0.0", port=80, debug=1)
